sapicore/engine/network/__init__.py

```diff
@@ -0,0 +1,400 @@
+
+        Note
+        ----
+        If your model requires some preprocessing of inputs to the postsynaptic neuron, it can be implemented
+        by overriding this method.
+
+        """
+        return torch.sum(torch.vstack(synaptic_input), dim=0)
+
+    def backward(self) -> None:
+        """Processes a backward sweep for this network object.
+
+        This is not required for SNNs that learn with STDP and can be implemented by child classes that
+        might require a backward pass.
+
+        """
+        pass
+
+    def forward(self, data: torch.tensor) -> dict:
+        """Processes current simulation step for this network object.
+
+        In this generic implementation, forward call order is determined by a BFS traversal starting from
+        the edges fanning out of the root nodes. Every node is visited once; for each node, every outgoing synapse
+        is forwarded. That way, we are guaranteed to forward any component exactly once on every simulation step.
+
+        Parameters
+        ----------
+        data: torch.tensor
+            External input to be processed by this generic network.
+
+        Returns
+        -------
+        dict
+            Dictionary containing the loggable properties of this network's ensembles and synapses in this
+            simulation step.
+
+        Warning
+        -------
+        Networks now support multiple root nodes and external input streams. In those cases, the graph is
+        traversed layer-wise, using multi-source BFS.
+
+        See Also
+        --------
+        :meth:`simulation.Simulator.run`
+            For a demonstration of how to programmatically feed heterogeneous external current to multiple nodes.
+            Note that this functionality will eventually be relegated to the data loader, with arbitrary current
+            injections during the simulation being implemented by biases.
+
+        """
+        # follow outgoing synapses between vertices, forwarding the entire network on each iteration.
+        for i, ensemble in enumerate(self.traversal_order):
+            # list all synapses coming in and out of this ensemble. Networkx references it by its string identifier.
+            incoming_synapses = self._in_edges(ensemble)
+            outgoing_synapses = self._out_edges(ensemble)
+
+            # shortcut to actual ensemble reference.
+            ensemble_ref = self.graph.nodes[ensemble].get("reference")
+
+            if ensemble_ref.identifier not in self.roots:
+                # apply a summation function to synaptic data flowing into this ensemble (torch.sum by default).
+                integrated_data = self.summation([synapse.output for synapse in incoming_synapses]).to(self.device)
+
+            else:
+                # if this is a root node, treat its stream from the data loader as inbound synaptic input.
+                external = [data[self.roots.index(ensemble_ref.identifier)]] if isinstance(data, list) else [data]
```

Comment:
I'm not sure what is happening here conceptually!? Can data be a list and not a tensor?

Reply:
Part of the accommodations made for multiple data streams (for networks with multiple roots). This should probably be clarified, but a user may pass a list of tensors (samples X features) where the i-th root receives the i-th external data tensor. Could be done with 3D tensors too, of course, but I'm sure something biased my decision when I made it.